File: Source/Model/MockUserClient.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


#import "MockUserClient+Internal.h"
#import "MockPreKey.h"


@implementation MockUserClient

@dynamic user;
@dynamic label;
@dynamic type;
@dynamic identifier;
@dynamic prekeys;
@dynamic lastPrekey;
@dynamic mackey;
@dynamic enckey;
@dynamic address;
@dynamic deviceClass;
@dynamic locationLatitude;
@dynamic locationLongitude;
@dynamic model;
@dynamic time;

@synthesize encryptionContext;

+ (NSFetchRequest *)fetchRequestWithPredicate:(NSPredicate *)predicate
{
    NSFetchRequest *request = [NSFetchRequest fetchRequestWithEntityName:@"UserClient"];
    request.predicate = predicate;
    return request;
}

+ (instancetype)insertClientWithPayload:(NSDictionary *)payload contenxt:(NSManagedObjectContext *)context
{
    NSString *type = [payload optionalStringForKey:@"type"];
    NSString *label = [payload optionalStringForKey:@"label"];
    NSDictionary *lastKeyPayload = [payload optionalDictionaryForKey:@"lastkey"];
    NSArray *prekyesPayload = [payload optionalArrayForKey:@"prekeys"];
    NSDictionary *sigkeysPayload = [payload optionalDictionaryForKey:@"sigkeys"];
    NSString *mackey = [sigkeysPayload optionalStringForKey:@"mackey"];
    NSString *enckey = [sigkeysPayload optionalStringForKey:@"enckey"];
    NSString *deviceClass = [payload optionalStringForKey:@"class"];
    NSString *model = [payload optionalStringForKey:@"model"];
    
    if (type == nil || !([type isEqualToString:@"temporary"] || [type isEqualToString:@"permanent"]) ||
        lastKeyPayload == nil || prekyesPayload == nil || mackey == nil || enckey == nil) {
        return nil;
    }
    
    if ([[lastKeyPayload numberForKey:@"id"] integerValue] != 0xFFFF) {
        return nil;
    }
    
    
    MockUserClient *newClient = [NSEntityDescription insertNewObjectForEntityForName:@"UserClient" inManagedObjectContext:context];
    newClient.label = label;
    newClient.type = type;
    newClient.identifier = [NSString createAlphanumericalString];
    newClient.mackey = mackey;
    newClient.enckey = enckey;
    newClient.deviceClass = deviceClass;
    newClient.model = model;
    newClient.locationLatitude = 52.5167;
    newClient.locationLongitude = 13.3833;
    newClient.address = @"62.96.148.44";
    newClient.time = [NSDate date];

    NSSet *prekeys = [MockPreKey insertNewKeysWithPayload:prekyesPayload context:context];
    for (MockPreKey *prekey in prekeys) {
        prekey.client = newClient;
    }
    MockPreKey *lastPreKey = [MockPreKey insertNewKeyWithPayload:lastKeyPayload context:context];
    lastPreKey.client = newClient;
    
    newClient.prekeys = prekeys;
    newClient.lastPrekey = lastPreKey;

    
    return newClient;
}

+ (instancetype)insertClientWithLabel:(NSString *)label type:(NSString *)type atLocation:(NSURL *)location inContext:(NSManagedObjectContext *)moc;
{
    MockUserClient *newClient = [NSEntityDescription insertNewObjectForEntityForName:@"UserClient" inManagedObjectContext:moc];

    newClient.identifier = [NSString createAlphanumericalString];
    newClient.label = label;
    newClient.type = type;
    
    newClient.time = [NSDate date];

    NSURL *clientOtrLocation = [location URLByAppendingPathComponent:[NSString stringWithFormat:@"%@_%@", newClient.identifier, newClient.label]];
    [[NSFileManager defaultManager] createDirectoryAtURL:clientOtrLocation withIntermediateDirectories:YES attributes:nil error:nil];
    EncryptionContext *encryptionContext = [[EncryptionContext alloc] initWithPath:clientOtrLocation];
    VerifyReturnNil(encryptionContext != nil);
    newClient.encryptionContext = encryptionContext;
    
    __block NSArray *prekeys;
    __block NSString *lastPrekey;
    __block NSError *error;
    [encryptionContext perform:^(EncryptionSessionsDirectory * _Nonnull sessionsDirectory) {
        prekeys = [sessionsDirectory generatePrekeys:NSMakeRange(0, 100) error:&error];
        lastPrekey = [sessionsDirectory generateLastPrekeyAndReturnError:&error];
    }];
    VerifyReturnNil([prekeys count] > 0 && error == nil);
    NSArray *preKeyStrings = [prekeys mapWithBlock:^id(NSDictionary *keyInfo) {
        return keyInfo[@"prekey"];
    }];
    VerifyReturnNil([prekeys count] > 0 && error == nil);
    
    NSArray <MockPreKey *> *mockPrekeys = [MockPreKey insertMockPrekeysFromPrekeys:preKeyStrings forClient:newClient inManagedObjectContext:moc];
    newClient.prekeys = [NSSet setWithArray:mockPrekeys];
    
    MockPreKey *mockLastPrekey = [MockPreKey insertNewKeyWithPrekey:lastPrekey forClient:newClient inManagedObjectContext:moc];
    newClient.lastPrekey = mockLastPrekey;
    
    VerifyReturnNil(prekeys);
    
    
    return newClient;
}

- (id<ZMTransportData>)transportData;
{
    NSMutableDictionary *data = [NSMutableDictionary dictionary];
    data[@"id"] = self.identifier;
    data[@"label"] = self.label ?: [NSNull null];
    data[@"type"] = self.type;
    data[@"time"] = self.time.transportString;
    if(self.model != nil) {
        data[@"model"] = self.model;
    }
    if(self.deviceClass != nil) {
        data[@"class"] = self.deviceClass;
    }
    data[@"address"] = self.address;
    data[@"location"] = @{
                           @"lat" : @(self.locationLatitude),
                           @"lon" : @(self.locationLongitude)
                           };
    return data;
}

+ (NSData *)encryptedDataFromClient:(MockUserClient *)fromClient toClient:(MockUserClient *)toClient data:(NSData *)data
{
    __block NSError *error;
    __block NSData *encryptedData;
    EncryptionContext *encryptionContext = fromClient.encryptionContext;
    [encryptionContext perform:^(EncryptionSessionsDirectory * _Nonnull sessionsDirectory) {
        if (![sessionsDirectory hasSessionForID:toClient.identifier]) {
            [sessionsDirectory createClientSession:toClient.identifier base64PreKeyString:toClient.lastPrekey.value error:&error];
        }
        encryptedData = [sessionsDirectory encrypt:data recipientClientId:toClient.identifier error:&error];
    }];
    return encryptedData;
}


+ (NSData *)sessionMessageDataForEncryptedDataFromClient:(MockUserClient *)fromClient toClient:(MockUserClient *)toClient data:(NSData *)data
{
    EncryptionContext *encryptionContext = toClient.encryptionContext;
    __block NSData *decryptedData;
    [encryptionContext perform:^(EncryptionSessionsDirectory * _Nonnull sessionsDirectory) {
        NSError *error;
        decryptedData = [sessionsDirectory createClientSessionAndReturnPlaintext:fromClient.identifier prekeyMessage:data error:&error];
    }];
    return decryptedData;
}

- (BOOL)establishConnectionWithClient:(MockUserClient *)client
{
    __block NSError *error = nil;
    __block BOOL hasSession = NO;
    [self.encryptionContext perform:^(EncryptionSessionsDirectory * _Nonnull sessionsDirectory) {
        if (! [sessionsDirectory hasSessionForID:client.identifier]) {
            [sessionsDirectory createClientSession:client.identifier base64PreKeyString:client.lastPrekey.value error:&error];
        }
        hasSession = [sessionsDirectory hasSessionForID:client.identifier];
    }];
    return hasSession;
}

- (void)dealloc
{
    self.encryptionContext = nil;
}


@end

@implementation NSString (RandomString)

+ (NSString *)createAlphanumericalString {
    u_int64_t number = 0;
    arc4random_buf(&number, sizeof(u_int64_t));
    NSString *string = [NSString stringWithFormat:@"%llx", number % LONG_LONG_MAX];
    return string;
}

@end

File: wire-ios-mocktransport/Source/Clients and OTR/MockUserClient.swift
//
// Wire
// Copyright (C) 2025 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import CoreData
import Foundation

@objc
public class MockUserClient: NSManagedObject {

    /// User that owns the client
    @NSManaged public var user: MockUser?

    /// Remote identifier
    @NSManaged public var identifier: String?

    /// Device label
    @NSManaged public var label: String?

    /// Device type
    @NSManaged public var type: String?

    /// IP address of registration
    @NSManaged public var address: String?

    /// Device class
    @NSManaged public var deviceClass: String?

    /// Registration location latitude
    @NSManaged public var locationLatitude: Double

    /// Registration location longitude
    @NSManaged public var locationLongitude: Double

    /// Device model
    @NSManaged public var model: String?

    /// Time of registration
    @NSManaged public var time: Date?

    /// Encryption key for APS signalling
    @NSManaged public var enckey: String?

    /// Hashing key for APS signalling
    @NSManaged public var mackey: String?

    /// Prekeys registered for this client
    @NSManaged public var prekeys: Set<MockPreKey>

    /// Last prekeys registered for this client
    @NSManaged public var lastPrekey: MockPreKey

}

public extension MockUserClient {

    /// Returns a fetch request to fetch MockUserClients with the given predicate
    @objc
    static func fetchRequest(predicate: NSPredicate) -> NSFetchRequest<MockUserClient> {
        let request = NSFetchRequest<MockUserClient>(entityName: "UserClient")
        request.predicate = predicate
        return request
    }
}

// MARK: - Legal Hold

public extension MockUserClient {

    var isLegalHoldDevice: Bool {
        type == "legalhold" || deviceClass == "legalhold"
    }

}

// MARK: - JSON de/serialization

@objc
public extension MockUserClient {

    /// Creates a new client from JSON payload
    static func insertClient(payload: [String: Any], context: NSManagedObjectContext) -> MockUserClient? {

        let label = payload["label"] as? String
        let deviceClass = payload["class"] as? String
        let model = payload["model"] as? String

        guard let type = payload["type"] as? String, validClientTypes.contains(type),
              let sigkeysPayload = payload["sigkeys"] as? [String: Any],
              let lastKeyPayload = payload["lastkey"] as? [String: Any],
              let prekeysPayload = payload["prekeys"] as? [[String: Any]],
              let mackey = sigkeysPayload["mackey"] as? String,
              let enckey = sigkeysPayload["enckey"] as? String,
              let prekeyNumber = lastKeyPayload["id"] as? Int, prekeyNumber == 0xFFFF
        else {
            return nil
        }

        let newClient = NSEntityDescription.insertNewObject(
            forEntityName: "UserClient",
            into: context
        ) as! MockUserClient
        newClient.label = label
        newClient.type = type
        newClient.identifier = String.randomClientIdentifier()
        newClient.mackey = mackey
        newClient.enckey = enckey
        newClient.deviceClass = deviceClass
        newClient.model = model
        newClient.locationLatitude = 52.5167
        newClient.locationLongitude = 13.3833
        newClient.address = "62.96.148.44"
        newClient.time = Date()

        let prekeys = MockPreKey.insertNewKeys(withPayload: prekeysPayload, context: context)
        prekeys.forEach {
            $0.client = newClient
        }

        let lastPreKey = MockPreKey.insertNewKey(withPayload: lastKeyPayload, context: context)!
        lastPreKey.client = newClient

        newClient.prekeys = prekeys
        newClient.lastPrekey = lastPreKey

        return newClient
    }

    /// Insert a new client, automatically generate prekeys and last key
    @objc(insertClientWithLabel:type:deviceClass:user:context:)
    static func insertClient(
        label: String,
        type: String = "permanent",
        deviceClass: String = "phone",
        for user: MockUser,
        in context: NSManagedObjectContext
    ) -> MockUserClient? {
        let newClient = NSEntityDescription.insertNewObject(
            forEntityName: "UserClient",
            into: context
        ) as! MockUserClient

        newClient.user = user
        newClient.identifier = String.randomClientIdentifier()
        newClient.label = label
        newClient.type = type
        newClient.deviceClass = deviceClass
        newClient.time = Date()

        // Generate mock prekey strings (no encryption needed for mock transport)
        let mockPrekeys = (0..<5).map { _ in UUID().uuidString }
        let mockLastPrekey = UUID().uuidString

        let prekeys = MockPreKey.insertNewKeys(
            withPayload: mockPrekeys,
            context: context
        )
        newClient.prekeys = Set(prekeys)

        let lastPrekey = MockPreKey.insertNewKey(withPrekey: mockLastPrekey, for: newClient, in: context)
        lastPrekey.identifier = Int(UInt16.max)
        newClient.lastPrekey = lastPrekey

        return newClient
    }

    /// JSON representation
    var transportData: ZMTransportData {

        var data = [String: Any]()
        data["id"] = identifier
        if label != nil {
            data["label"] = label
        }
        data["type"] = type
        if let time {
            data["time"] = time.transportString()
        }
        if let model {
            data["model"] = model
        }
        if let device = deviceClass {
            data["class"] = device
        }
        data["address"] = address
        data["location"] = [
            "lat": locationLatitude,
            "lon": locationLongitude
        ]
        return data as NSDictionary
    }
}

/// Allowed client types
private let validClientTypes = Set(["temporary", "permanent", "legalhold"])
